core/server/api/themes.js

```diff
@@ -26,23 +23,41 @@ var debug = require('debug')('ghost:api:themes'),
  */
 themes = {
     browse: function browse() {
-        debug('browsing');
-        var result = themeList.toAPI(themeList.getAll(), settingsCache.get('activeTheme'));
-        debug('got result');
-        return Promise.resolve({themes: result});
+        return Promise.resolve(themeUtils.toJSON());
     },
 
     activate: function activate(options) {
         var themeName = options.name,
             newSettings = [{
                 key: 'activeTheme',
                 value: themeName
-            }];
+            }],
+            loadedTheme,
+            checkedTheme;
+
+        return apiUtils
+            .handlePermissions('themes', 'activate')(options)
+            .then(function activateTheme() {
+                loadedTheme = themeList.get(themeName);
+
+                if (!loadedTheme) {
+                    return Promise.reject(new errors.ValidationError({
+                        message: i18n.t('notices.data.validation.index.themeCannotBeActivated', {themeName: themeName}),
+                        context: 'activeTheme'
+                    }));
+                }
 
-        return settings.edit({settings: newSettings}, options).then(function () {
-            var result = themeList.toAPI(themeList.getAll(), themeName);
-            return Promise.resolve({themes: result});
-        });
+                return themeUtils.validate.check(loadedTheme);
+            })
+            .then(function haveValidTheme(_checkedTheme) {
+                checkedTheme = _checkedTheme;
+                // We use the model, not the API here, as we don't want to trigger permissions
+                return settingsModel.edit(newSettings, options);
+            })
+            .then(function hasEditedSetting() {
+                // @TODO actually do things to activate the theme, other than just the setting?
```

Comment:
This is a foreshadowing of my next PR 😉  in which I properly introduce a concept of "activating" a theme.

Atm we have 3 concepts:

1. on boot, sort of activate a theme by doing some loading
2. on activate, sort of activate a theme by changing a setting
3. on upload, sort of activate a theme by overriding and having a [dirty hack](https://github.com/TryGhost/Ghost/blob/master/core/server/api/index.js#L69).

In all 3 cases, the next request magically "fixes" the state here: https://github.com/TryGhost/Ghost/blob/master/core/server/middleware/theme-handler.js

All of which means in the current setup I cannot implement any of the features we need properly without hacks.

Next PR will introduce a better concept of both "activating" and then "mounting" a theme, from which the new features become simple changes.